src/Capability/Registry.php

```diff
@@ -338,6 +531,156 @@ public function getPrompt(string $name): PromptReference
         return $this->prompts[$name] ?? throw new PromptNotFoundException($name);
     }
 
+    public function getDynamicToolProviders(): array
+    {
+        return $this->dynamicToolProviders;
+    }
+
+    public function getDynamicPromptProviders(): array
+    {
+        return $this->dynamicPromptProviders;
+    }
+
+    public function getDynamicResourceProviders(): array
+    {
+        return $this->dynamicResourceProviders;
+    }
+
+    /**
+     * @return array<DynamicResourceTemplateProviderInterface>
+     */
+    public function getDynamicResourceTemplateProviders(): array
+    {
+        return $this->dynamicResourceTemplateProviders;
+    }
+
+    public function getDynamicPromptCompletionProviders(string $name): ?array
+    {
+        $provider = $this->findDynamicPromptProvider($name);
+
+        return $provider?->getCompletionProviders($name);
+    }
+
+    public function getDynamicResourceTemplateCompletionProviders(string $uri): ?array
+    {
+        $provider = $this->findDynamicResourceTemplateProvider($uri);
+
+        return $provider?->getCompletionProviders($uri);
+    }
+
+    private function findDynamicToolProvider(string $toolName): ?DynamicToolProviderInterface
+    {
+        $matching = array_filter(
+            $this->dynamicToolProviders,
+            fn (DynamicToolProviderInterface $p) => $p->supportsTool($toolName),
+        );
+
+        return array_shift($matching);
```

Comment:
Why only the first one? Why not return `iterable<DynamicToolProviderInterface>`?